File: 지찬/프로그래머스 코테 kit/주식가격.java
//https://school.programmers.co.kr/learn/courses/30/lessons/42584
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        Stack<Integer> stack = new Stack<>();
        int[] answer = new int[prices.length];
        for (int i = 0; i < prices.length; i++) {
            if (stack.isEmpty()) {
                stack.push(i);
            } else {
                while (!stack.isEmpty() && prices[stack.peek()] > prices[i]) {
                    int idx = stack.pop();
                    answer[idx] = i - idx;
                }
                stack.push(i);
            }
        }
        while (!stack.isEmpty()) {
            int idx = stack.pop();
            answer[idx] = prices.length - 1 - idx;
        }
        return answer;
    }


    public static void main(String[] args) {
        int[] prices = {
                1, 2, 3, 2, 3
        };
        Solution s = new Solution();
        System.out.println(Arrays.toString(s.solution(prices)));
    }
}

File: 종우/주식가격.py
# 0128
# https://school.programmers.co.kr/learn/courses/30/lessons/42584

# 내 풀이 효율성 x
def solution(prices):
    answer = [0]*len(prices)#[6,5,1,3,1,0,0,0]
    dropped = [0] #[0,1,4,5]
    for i in range(1,len(prices)):
        for j in dropped:
            if prices[i] >= prices[j]:
                answer[j] += 1
            else:
                answer[j] += 1
                dropped.remove(j)
        dropped.append(i)
    return answer

# 다른 풀이
def solution(prices):
    length = len(prices)
    answer = [ i for i in range (length - 1, -1, -1)]
    
    stack = [0]
    for i in range (1, length, 1):
        while stack and prices[stack[-1]] > prices[i]:
            j = stack.pop()
            answer[j] = i - j
        stack.append(i)
    return answer